Update rules (2023.07)

phanect opened the pull request:
いくつかルールを無効にしました。
無効にしたい理由については以下の通りです:

# `@typescript-eslint/no-misused-promises` (`checksVoidReturn` オプション)

`checksVoidReturn: true` となっている場合 (現状)、

```ts
window.addEventListener("load", async () => { /* ... */ })
```

のようなコードはエラーが出ます。`addEventListener` の第2引数は、`void` を返す関数を指定する必要があるのであって、`Promise<void>` を返す `async` 関数は想定されていないためです。

`checksVoidReturn: false` とすることで、このようなコードが許可されるようになります。

# `import/no-extraneous-dependencies`

現状、例えば webpack.config.js など、設定用のファイルで `devDependencies` からパッケージを `import` / `require` した際に警告が出ます。

例えば下記のようなコードを webpack.config.js から呼び出すと警告が出ます。

```js
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');
```

webpack-bundle-analyzer はビルドタイムにのみ必要なので、`devDependencies` にインストールされているべきですが、このルールは `dependencies` に移すようにと警告します。

----------

その他、ES2015 のソースコードを前提に lint する設定になっていたので、ES2024 まで対象とするよう ES バージョンを引き上げました。

phanect committed:
feat: disable `checksVoidReturn`

If `checksVoidReturn: true`, you cannot write code like follows:

```
window.addEventListener("load", async () => { /* ... */ })
```

because the second argument of `addEventListener` is a function returning `void`, not returning `Promise<void>`.

By disabling `checksVoidReturn`, you can assign async functions as arguments which is expected non-async functions.

phanect committed:
feat: set ES version to 2024

`parserOptions.ecmaVersion` is automatically set to 15 (== 2024) by setting `env.es2024: true`

kamataryo commented:
なるほど..
`import/no-extraneous-dependencies` ルールについてなのですが、 確か Serverless Framework でデプロイパッケージを作成すると、dependencies だけがパッキングされて、必要なライブラリを間違って devDependencies に入れていると Lambda 上で参照できなくなってしまうと思うので、個人的には `import/no-extraneous-dependencies` は欲しい（確かそのために入れたような？）感じです。

`webpack.config.js` を無視するような設定って可能でしょうか？

phanect committed:
chore: update yarn.lock

phanect commented:
@kamataryo 了解しました。
webpack.config.js など設定ファイル無視できないか確認してみます。

phanect committed:
feat: disable `import/no-extraneous-dependencies` for known config files

This rule warns imports from config files such as webpack.config.js.
For example, consider you have following code in webpack.config.js.

```js
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');
```

Since webpack-bundle-analyzer is only required at build time and it should be `devDependencies`, However, this rule requests to move it `dependencies`.

phanect commented:
@kamataryo [feat: disable import/no-extraneous-dependencies for known config files](https://github.com/geolonia/eslint-config/pull/8/commits/c0f2fa14ae7dfa1e11bf408500235c20c8082d8c) で対応しました。

kamataryo reviewed (approved):
ありがとうございます！
開発環境用の設定ファイルなど、必要になったら順次足していきましょう 👍
